### ISSUE TITLE
Do users always have to copy the message when writing to a non blocking TcpStream?

### ISSUE THREAD

**Issue by mvtec-bergdolll** [snapview/tungstenite-rs#281]
First of all, thank you for creating and maintaining this project! I have a situation where I share access to a WebSocket in a single client scenario, thus no tokio-tungstenite, but there might be multiple threads reading and writing from the same WebSocket connection, requiring protection via Mutex. Roughly in code:

```rust
// --- Setup ---
fn accept_connection(stream: TcpStream, ...) {

let websocket = accept(stream).expect("Failed WebSocket handshake");

websocket
    .get_ref()
    .set_nonblocking(true)
    .expect("Failed to set the stream to non blocking");

// --- Read ---
loop {
    let msg_opt = shared_websocket.lock().unwrap().read_message().no_block()?;

    [...]
}

// --- Write ---
fn send(&self, response: Vec<u8>) -> Result<(), ResponseError> {
    const MAX_SPIN_COUNT: u32 = 10;

    let mut spin_attempts = 0;

    while let None = self
        .shared_websocket
        .lock()
        .unwrap()
        .write_message(Message::Binary(response.clone()))
        .no_block()?
    {
        spin_attempts += 1;

        if spin_attempts == MAX_SPIN_COUNT {
            thread::sleep(Duration::from_millis(1));
            spin_attempts = 0;
        }
    }

    Ok(())
}
```

Maybe I misunderstood the implication of `set_nonblocking(true)`, my understanding is that it allows syscalls to return WOULDBLOCK which could also happen when one attempts to write. But the current design implies that we have to copy in the response message for each attempt. Which is unfortunate and doesn't seem to happen with the async interface. So I wonder if my understanding is wrong, or maybe the interface is not well suited for this use case.

**Comment by application-developer-DA** [comment]
The implication of making the socket non-blocking is that whenever we can't write/read to/from the socket, we return the `WOULDBLOCK`.

You actually do not need to copy the message multiple times even if you use the sockets in a non-blocking mode (`tungstenite` is agnostic to the type of the socket). If you're able to call the `write_message` and does not get the error message with your message returned back (due to the send queue being full), then it means that the message has been added to the queue, so you don't need to call the `write_message` with the same message again (unless you want to send a duplicate).

So after you've got the successful write with `write_message()` (even if the message was not fully sent due to the `WOULDBLOCK`), you should continue calling [`write_pending()`](https://docs.rs/tungstenite/latest/tungstenite/protocol/struct.WebSocket.html#method.write_pending) that will ensure that the send queue is flushed.

In other words:
- `write_message()` adds the message to the queue and starts sending it. If we were not able to send the message fully (flush the buffer), the `WOULDBLOCK` is returned (the "not yet sent" part of the message remains in the sending queue).
- `write_pending()` flushes the sending queue and succeeds when the queue is empty (i.e. when all "scheduled" messages have been sent).

Hope this answers the question 🙂

**Comment by mvtec-bergdolll** [comment]
Yes, thank you!
